src/vm/i386/gmsx86.cpp

```diff
@@ -1295,10 +1295,10 @@ void LazyMachState::unwindLazyState(LazyMachState* baseState,
     ctx.Esi = lazyState->_esi = baseState->_esi;
     ctx.Ebx = lazyState->_ebx = baseState->_ebx;
 
-    nonVolRegPtrs.Edi = &(lazyState->_edi);
-    nonVolRegPtrs.Esi = &(lazyState->_esi);
-    nonVolRegPtrs.Ebx = &(lazyState->_ebx);
-    nonVolRegPtrs.Ebp = &(lazyState->_ebp);
+    nonVolRegPtrs.Edi = &(baseState->_edi);
```

Comment:
Note that ``HelperMethodFrame::InsureInit`` passes a pointer to a local variable as ``lazyState``, which means that ``&(lazyState->...)`` becomes invalid after ``HelperMethodFrame::InsureInit`` returns. 

``PAL_VirtualUnwind`` will not update this pointer if ``edi`` is not saved in the corresponding helper function. In this case, `` nonVolRegPtrs.Edi`` remains as ``&(lazyState->_edi)``, and thus ``LazyMachState::unwindLazyState`` sets ``lazyState->_pEdi`` as ``&(lazyState->_edi)`` just before returning.

EH system dereferences this invalid pointer to update a resume context if a managed exception is thrown inside any helper function that does not save ``edi``, and the managed code that invokes this helper function catches this exception.

This results in test failures mentioned in the PR description.